src/psyclone/psyGen.py

```diff
@@ -4250,6 +4266,188 @@ def nodes_to_code_block(parent, statements):
+                                        Fortran2003.Execution_Part)
+        except ValueError:
+            pass
+        else:
+            self.process_nodes(new_schedule, sub_exec.content, sub_exec)
+
+        return new_schedule
+
+    def process_declarations(self, parent, nodes, arg_list):
+        '''
+        Transform the variable declarations in fparser2 parse tree into
+        symbols in the PSyIR parent node symbol table.
+
+        :param parent: PSyIR node in which to insert the symbols found.
+        :type parent: :py:class:`psyclone.psyGen.KernelSchedule`
+        :param nodes: fparser2 AST nodes to search for declaration statements.
+        :type nodes: list of :py:class:`fparser.two.utils.Base`
+        :param arg_list: fparser2 AST node containing the argument list.
+        :type arg_list: :py:class:`fparser.Fortran2003.Dummy_Arg_List`
+        :raises NotImplementedError: The provided declarations contain
+                                     attributes which are not supported yet.
+        '''
+        from fparser.two.utils import walk_ast
+        from fparser.two import Fortran2003
+        for decl in walk_ast(nodes, [Fortran2003.Type_Declaration_Stmt]):
+            # Parse data type, currently restricted to integers and reals.
+            datatype = 'unknown'
+            if str(decl.items[0]).lower().startswith('real'):
+                datatype = 'real'
+            elif str(decl.items[0]).lower().startswith('integer'):
+                datatype = 'integer'
+
+            # Parse number of dimensions if it is an array
+            shape = []
+
+            for attr in walk_ast(decl.items,
+                                 [Fortran2003.Assumed_Shape_Spec,
+                                  Fortran2003.Explicit_Shape_Spec,
+                                  Fortran2003.Assumed_Size_Spec
+                                  ]):
+                if isinstance(attr, Fortran2003.Assumed_Size_Spec):
+                    raise NotImplementedError("Could not process {0}. Assumed"
+                                              "-size arrays are not "
+                                              "supported.".format(nodes))
+                elif isinstance(attr, Fortran2003.Assumed_Shape_Spec):
+                    shape.append(None)
+                elif isinstance(attr, Fortran2003.Explicit_Shape_Spec):
+                    if isinstance(attr.items[1],
+                                  Fortran2003.Int_Literal_Constant):
+                        shape.append(int(attr.items[1].items[0]))
+                    else:
+                        raise NotImplementedError("Could not process {0}. "
+                                                  "Only integer literals are "
+                                                  "supported for explicit "
+                                                  "shape array declarations."
+                                                  "".format(nodes))
```

Comment:
Please add an `else InternalError` to defend ourselves against a future developer modifying the call to `walk_ast` but then forgetting to update this if block. I expect you'll then have to monkeypatch `walk_ast` in order to get a test to cover this branch of the if block.

Reply:
Done, I have moved the parsing of dimension attributes to a static private method, so it is easy to monkeypatch just this specific `walk_ast`.